Dark theme using CSS variables and local storage (#1700)
section-engineering-education/engineering-education: articles/adding-dark-theme-to-your-site/index.md

The dark theme has gained prevalence in screens today. With this trend in iOS, macOS, windows, and google, most systems adopt dark themes. Application of dark theme in website context make your website more exciting and attractive to users who love dark them. The experience is even better if you include functionality to switch between light and dark themes. This article will help you learn how to implement switching between light and dark themes using [CSS Varibles](https://www.w3schools.com/css/css3_variables.asp). Click [this](https://replit.com/@PhinaKersly/dark-theme#index.html) link to find the source code and a runnable program for the implememntation.

> Before we begin, we will need the following:
> ### Prerequisites
> 1. Your favorite text editor. I Prefer vscode
> 2. Basic Knowledge of HTML, SCSS, and Javascript
> 3. Understand CSS variables

Reviewer: ```suggestion
### Prerequisites
As a prerequisite, the reader must have a good understanding of:
- Basic Knowledge of HTML, SCSS, and Javascript
- Understand CSS variables
```

> 

Reviewer: Can you add a `Table of contents` section? Something like [this](https://www.section.io/engineering-education/introduction-to-adafruit-io/#table-of-contents)

```html
<!DOCTYPE html>
<html lang="en" data-theme="light">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="stylesheet" href="css/main.css">
    <title>Document</title>
</head>
<body>
    <div class="container">
        <h1>Section Engineering Education</h1>
        <div class="toggle-container">
            <input type="checkbox" id="switch" name="theme" /><label for="switch">Toggle</label>
        </div>

        <p>Section partners with university students in Computer Science related fields of study to research and write about topics that are relevant to engineers in the modern technology landscape. You can find more information and program guidelines in the GitHub repository. If you're currently enrolled in a Computer Science related field of study and are interested in participating in the program, please complete <a href="https://docs.google.com/forms/d/e/1FAIpQLSfTbj3kqvEJEb5RLjqJurfbHa8ckzQx0CjRzaizblue9ZOK5A/viewform">this form </a></p>
    </div>
    <script src="main.js"></script>
</body>
</html>
```

> 

Reviewer: Add an image for displaying the HTML output

### Adding the CSS
In the same directory as the HTML file, we will add the CSS file that will be used for toggling the default white theme to the dark theme. I recommend installing a live SCSS Compiler to compile our CSS Code in realtime.
The CSS variables for the default light theme are as shown below. You can change the color code to your favorite color and look attractive in both dark and light themes.

The CSS styling for the default white theme is shown below.

> ```CSS

Reviewer: ```suggestion
```css
```

html{
    --bg: #fff;
    --bg-panel: #ebebeb;
    --color-heading: rgb(27, 168, 14); 
    --color-text: #333333;
}
```

CSS styling for dark theme.

> ```CSS

Reviewer: ```suggestion
```css
```

html[data-theme='dark'] {
    --bg: #333333;
    --bg-panel: #434343;
    --color-heading: #0077ff; 
    --color-text: #B5B5B5;
}
```

Next, we need to specify the CSS variables' use through the CSS file so that all the website elements will change when we click the toggle theme button.

```css
body { 
    background-color: var(--bg);
}

.container {
    background-color: var(--bg-panel);

    h1{
        margin: 0;
        color: var(--color-heading)
    }

    p{
        color: var(--color-text);
       
    }

}
```

> 

Reviewer: Add an image after adding CSS to the HTML

### Switch Theme Button
In the next section, we will style our toggle switch which will be cheched to switch bewten the dark and ligh theme. The code for this toggle button cis as below. You can find the  code to this switch [here](https://codepen.io/mburnette/pen/LxNxNg)

```css
    input[type=checkbox] {
        height: 0;
        width: 0;
        visibility: hidden;
    }

    label {
        cursor: pointer;
        text-indent: -9999px;
        width: 52px;
        height: 27px;
        background: #1ba80e;
        float: right;
        border-radius: 100px;
        position: relative;
    }

    label:after {
        content: '';
        position: absolute;
        top: 3px;
        left: 3px;
        width: 20px;
        height: 20px;
        background: #fff;
        border-radius: 90px;
        -webkit-transition: 0.3s;
        transition: 0.3s;
    }

    input:checked + label {
        background: var(--color-heading);
    }

    input:checked + label:after {
        left: calc(100% - 5px);
        -webkit-transform: translateX(-100%);
                transform: translateX(-100%);
    }

    label:active:after {
        width: 45px;
    }
```

> 

Reviewer: Add an image displaying the toggle button as shown in Codepen.

### Adding the JavaScript
We will handle the theme switching through JavaScript by changing the toggle switch's class name to either light or dark, as defined above. In doing so, we write two functions `changeThemeToDark()` and `chnageThemeToWhite()`.

**Change the theme to dark.**
```javascript
    // Change theme to dark by adding the `dark` classname to html element.
   const changeThemeToDark = () =>{
    document.documentElement.setAttribute("data-theme", "dark")
}

```

**Change the theme to Light.**
```javascript
    // Reset the html class to default
   const changeThemeToDark = () =>{
    document.documentElement.setAttribute("data-theme", "dark");
}
```

> ### Using Local Storage to remember the theme.

Reviewer: If possible, add 2 - 3 lines explaining what local storage is. (Technical aspects)

> Now that our dark theme is set, we need to enable the browser to remember the theme our user chose so that in subsequent visits or page refreshing, they see their preferred theme. We can use local storage to implement this functionality. We will store the applied theme in local storage and read subsequent calls of the `changeThemeToDark()` and `chnageThemeToWhite().` Even after refreshing the page, the theme is still reserved. We are using local storage because the data stored in local storage has no expiration time. The 

Reviewer: ```suggestion
Now that our dark theme is set, we need to enable the browser to remember the theme our user chose, so that in subsequent visits or page refreshing, they see their preferred theme.

We can use local storage to implement this functionality. We will store the applied theme in local storage and read subsequent calls of the `changeThemeToDark()` and `changeThemeToWhite().` Even after refreshing the page, the theme is still reserved. We are using local storage because the data stored in local storage has no expiration time.
```

The below pieces of code are used to save and retrieve the theme from local storage:

> ```javascript

Reviewer: Add comments to all the below snippets. (For every line)

    const changeThemeToDark = () =>{
        document.documentElement.setAttribute("data-theme", "dark")
        localStorage.setItem("data-theme", "dark")// save theme to local storage
    }

    const changeThemeToLight = () =>{
        document.documentElement.setAttribute("data-theme", "light")
        localStorage.setItem("data-theme", 'light')// save theme to local storage
    }
```
After writing, the functions, we check to see what theme is set currently and toggle it.
```javascript
    // Apply retrived them to the website
    checkbox.addEventListener('change', ()=> {
        let theme = localStorage.getItem('data-theme'); // Retrieve saved them from local storage
        if (theme ==='dark'){
            changeThemeToLight()
        }else{
            changeThemeToDark()
        }   
    });
```
### Conclusion

> To sum up, since dark themes are trendy recently, this article explained how to implement dark themes and light themes switching using CSS variables and Javascript's LocalStorage. You can now try out the dark theme on your website following this article. You can find the link to the code used in the article [here](https://replit.com/@PhinaKersly/dark-theme#index.html)

Reviewer: ```suggestion
To sum up, since dark themes are trendy recently, this article explained how to implement dark themes and light themes switching using CSS variables and Javascript's localStorage.

You can now try out the dark theme on your website following this article. You can find the link to the code used in the article [here](https://replit.com/@PhinaKersly/dark-theme#index.html)
```
